src/irmin-pack/unix/control_file_intf.ml

```diff
@@ -0,0 +1,164 @@
+(*
+ * Copyright (c) 2022-2022 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+open! Import
+
+module Payload_v3 = struct
+  type from_v1_v2_post_upgrade = { entry_offset_at_upgrade_to_v3 : int63 }
+  [@@deriving irmin]
+  (** [entry_offset_at_upgrade_to_v3] is the offset of the first entry that is
+      known to have been created using [irmin_pack_version = `V2] or more. The
+      entries before that point may be v1 entries. V1 entries need an entry in
+      index because it is the only place their lenght is stored. *)
```

Comment:
I don't understand precisely what V1, V2 and V3 are (is there documentation somewhere I can look at?). I understand that v1 entries did not have a length stored with them; v2 entries always have a length stored with them; and v3 is the current version we are working towards with the layered store? 

Then "V1 entries need an entry in the index because..." - are you proposing that the index will be non-minimal (i.e., contain entries for objects that are not commits), even when layers/GC operates? I don't think this can work - layers/GC requires a minimal index (perhaps I am missing something... please let me know if I have misunderstood).

Reply:
Also, how are we reconciling that V1 entries need an entry in the index, when layers needs minimal indexing?